airflow/kubernetes/pod_generator.py

```diff
@@ -62,75 +68,116 @@ class PodGenerator:
+    :type path_or_string: Optional[str]
+    :param extract_xcom: Whether to bring up a container for xcom
+    :type extract_xcom: bool
     """
     def __init__(  # pylint: disable=too-many-arguments,too-many-locals
         self,
-        image,
-        name=None,
-        namespace=None,
-        volume_mounts=None,
-        envs=None,
-        cmds=None,
-        args=None,
-        labels=None,
-        node_selectors=None,
-        ports=None,
-        volumes=None,
-        image_pull_policy='IfNotPresent',
-        restart_policy='Never',
-        image_pull_secrets=None,
-        init_containers=None,
-        service_account_name=None,
-        resources=None,
-        annotations=None,
-        affinity=None,
-        hostnetwork=False,
-        tolerations=None,
-        security_context=None,
-        configmaps=None,
-        dnspolicy=None,
-        pod=None,
-        extract_xcom=False,
+        image: Optional[str] = None,
+        name: Optional[str] = None,
+        namespace: Optional[str] = None,
+        volume_mounts: Optional[List[Union[k8s.V1VolumeMount, dict]]] = None,
+        envs: Optional[Dict[str, str]] = None,
+        cmds: Optional[List[str]] = None,
+        args: Optional[List[str]] = None,
+        labels: Optional[Dict[str, str]] = None,
+        node_selectors: Optional[Dict[str, str]] = None,
+        ports: Optional[List[Union[k8s.V1ContainerPort, dict]]] = None,
+        volumes: Optional[List[Union[k8s.V1Volume, dict]]] = None,
+        image_pull_policy: str = 'IfNotPresent',
+        restart_policy: str = 'Never',
+        image_pull_secrets: Optional[str] = None,
+        init_containers: Optional[List[k8s.V1Container]] = None,
+        service_account_name: Optional[str] = None,
+        resources: Optional[Union[k8s.V1ResourceRequirements, dict]] = None,
+        annotations: Optional[Dict[str, str]] = None,
+        affinity: Optional[dict] = None,
+        hostnetwork: bool = False,
+        tolerations: Optional[list] = None,
+        security_context: Optional[Union[k8s.V1PodSecurityContext, dict]] = None,
+        configmaps: Optional[List[str]] = None,
+        dnspolicy: Optional[str] = None,
+        pod: Optional[k8s.V1Pod] = None,
+        path_or_string: Optional[str] = None,
+        extract_xcom: bool = False,
     ):
-        self.ud_pod = pod
+        if pod and path_or_string:
+            raise AirflowConfigException("Cannot pass both `pod` and `path_or_string` arguments")
+
+        if path_or_string:
+            self.ud_pod = self.deserialize_model(path_or_string)
+        else:
+            self.ud_pod = pod
+
         self.pod = k8s.V1Pod()
         self.pod.api_version = 'v1'
         self.pod.kind = 'Pod'
 
         # Pod Metadata
         self.metadata = k8s.V1ObjectMeta()
         self.metadata.labels = labels
-        self.metadata.name = name + "-" + str(uuid.uuid4())[:8] if name else None
```

Comment:
What is the reason for this change?

Reply:
The `KubernetesExecutor` also [generates UUIDs](https://github.com/apache/airflow/blob/03eede00c90aa189d000776e3a263fdda6905cda/airflow/executors/kubernetes_executor.py#L496) for the pod name. I found that for pods generated by the `KubernetesExecutor` there were in fact two UUIDs, one from the Executor and one from the `PodGenerator` class itself. I wrongly assumed that in the case of the `KubernetesPodOperator` that the caller could enforce uniqueness of naming. You rightly pointed out a scenario I missed.
> Without this, backfilling will not work. 
My question would be, should the UUID generation exist inside the `PodGenerator` class, or push it to users of the `PodGenerator` class, so that the UUID generation would exist in the `KubernetesPodOperator` and remain as it is in the `KubernetesExecutor`.

Reply:
I've moved uuid generation into inside the `PodGenerator`. This simplifies the `AirflowKubernetesScheduler` as it is no longer responsible for the task. Technically I could move all field validation inside `PodGenerator`, which would make more sense to me.